Source/Scene/ModelInstanceCollection.js

```diff
@@ -0,0 +1,671 @@
+        '../Core/RuntimeError',
+        '../Renderer/Buffer',
+        '../Renderer/BufferUsage',
+        '../Renderer/DrawCommand',
+        '../Renderer/ShaderSource',
+        '../ThirdParty/when',
+        './Model',
+        './SceneMode'
+    ], function(
+        BoundingSphere,
+        Cartesian3,
+        clone,
+        Color,
+        combine,
+        ComponentDatatype,
+        defaultValue,
+        defined,
+        defineProperties,
+        DeveloperError,
+        destroyObject,
+        Matrix4,
+        RuntimeError,
+        Buffer,
+        BufferUsage,
+        DrawCommand,
+        ShaderSource,
+        when,
+        Model,
+        SceneMode) {
+    "use strict";
+
+    var LoadState = {
+        NEEDS_LOAD : 0,
+        LOADING : 1,
+        LOADED : 2,
+        FAILED : 3
+    };
+
+    /**
+     * A 3D model instance collection. All instances reference the same underlying model, but have unique
+     * per-instance properties like model matrix, pick id, etc.
+     *
+     * @alias ModelInstanceCollection
+     * @constructor
+     *
+     * @param {Object} options Object with the following properties:
+     * @param {String} options.url The url to the .gltf file.
+     * @param {Boolean} [options.dynamic] Collection is set to stream instance data every frame.
+     * @param {Object} [options.headers] HTTP headers to send with the request.
+     * @param {Boolean} [options.show=true] Determines if the collection will be shown.
+     * @param {Boolean} [options.allowPicking=false] When <code>true</code>, each glTF mesh and primitive is pickable with {@link Scene#pick}.
+     * @param {Boolean} [options.asynchronous=true] Determines if model WebGL resource creation will be spread out over several frames or block until completion once all glTF files are loaded.
+     * @param {Boolean} [options.debugShowBoundingVolume=false] For debugging only. Draws the bounding sphere for the collection.
+     * @param {Boolean} [options.debugWireframe=false] For debugging only. Draws the instances in wireframe.
+     *
+     * @private
+     */
+    var ModelInstanceCollection = function(options) {
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+
+        //>>includeStart('debug', pragmas.debug);
+        if (!defined(options.url)) {
+            throw new DeveloperError('options.url is required.');
+        }
+        //>>includeEnd('debug');
+
+        this._instances = defaultValue(options.instances, []);
+        this._instancingSupported = false;
+        this._dynamic = defaultValue(options.dynamic, false);
+        this._show = options.show;
+        this._allowPicking = defaultValue(options.allowPicking, false);
+        this._ready = false;
+        this._readyPromise = when.defer();
+        this._state = LoadState.NEEDS_LOAD;
+
+        this._model = undefined;
+        this._typedArray = undefined;
+        this._vertexBuffer = undefined;
+        this._createVertexBuffer = true;
```

Comment:
I added this so in the future it's easy to recreate the vertex buffer when instances are added by setting this flag.